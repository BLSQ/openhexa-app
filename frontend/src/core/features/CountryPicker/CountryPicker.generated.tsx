import * as Types from '../../../graphql/types';

import { gql } from '@apollo/client';
export type CountryPicker_CountryFragment = { __typename?: 'Country', code: string, alpha3: string, name: string };

export const CountryPicker_CountryFragmentDoc = gql`
    fragment CountryPicker_country on Country {
  code
  alpha3
  name
}
    `;