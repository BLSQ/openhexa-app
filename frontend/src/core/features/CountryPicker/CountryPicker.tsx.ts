import { graphql } from "graphql/gql";

export const CountryPickerCountryDoc = graphql(`
fragment CountryPicker_country on Country {
  code
  alpha3
  name
}
`);
