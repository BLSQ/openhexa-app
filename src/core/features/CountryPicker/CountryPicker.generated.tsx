import * as Types from '../../../graphql-types';

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
const defaultOptions = {} as const;
export type CountryPickerQueryVariables = Types.Exact<{ [key: string]: never; }>;


export type CountryPickerQuery = { __typename?: 'Query', countries: Array<{ __typename?: 'Country', code: string, alpha3: string, name: string, flag: string }> };

export type CountryPicker_CountryFragment = { __typename?: 'Country', code: string, alpha3: string, name: string, flag: string };

export const CountryPicker_CountryFragmentDoc = gql`
    fragment CountryPicker_country on Country {
  code
  alpha3
  name
  flag
}
    `;
export const CountryPickerDocument = gql`
    query CountryPicker {
  countries {
    ...CountryPicker_country
  }
}
    ${CountryPicker_CountryFragmentDoc}`;

/**
 * __useCountryPickerQuery__
 *
 * To run a query within a React component, call `useCountryPickerQuery` and pass it any options that fit your needs.
 * When your component renders, `useCountryPickerQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useCountryPickerQuery({
 *   variables: {
 *   },
 * });
 */
export function useCountryPickerQuery(baseOptions?: Apollo.QueryHookOptions<CountryPickerQuery, CountryPickerQueryVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useQuery<CountryPickerQuery, CountryPickerQueryVariables>(CountryPickerDocument, options);
      }
export function useCountryPickerLazyQuery(baseOptions?: Apollo.LazyQueryHookOptions<CountryPickerQuery, CountryPickerQueryVariables>) {
          const options = {...defaultOptions, ...baseOptions}
          return Apollo.useLazyQuery<CountryPickerQuery, CountryPickerQueryVariables>(CountryPickerDocument, options);
        }
export type CountryPickerQueryHookResult = ReturnType<typeof useCountryPickerQuery>;
export type CountryPickerLazyQueryHookResult = ReturnType<typeof useCountryPickerLazyQuery>;
export type CountryPickerQueryResult = Apollo.QueryResult<CountryPickerQuery, CountryPickerQueryVariables>;