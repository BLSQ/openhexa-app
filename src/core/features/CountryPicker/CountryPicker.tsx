import { gql, useQuery } from "@apollo/client";
import Combobox from "core/components/forms/Combobox";
import { ensureArray } from "core/helpers/array";
import useDebounce from "core/hooks/useDebounce";
import { useTranslation } from "next-i18next";
import { useCallback, useMemo, useState } from "react";
import {
  CountryPickerQuery,
  CountryPickerQueryVariables,
  CountryPicker_CountryFragment,
} from "./CountryPicker.generated";

type CountryPickerProps = {
  disabled?: boolean;
  placeholder?: string;
  required?: boolean;
  withPortal?: boolean;
  multiple?: boolean;
  value: CountryPicker_CountryFragment | CountryPicker_CountryFragment[] | null;
  onChange: (
    value:
      | CountryPicker_CountryFragment
      | CountryPicker_CountryFragment[]
      | null
  ) => void;
};

const CountryPicker = (props: CountryPickerProps) => {
  const { t } = useTranslation();
  const {
    value,
    onChange,
    disabled = false,
    multiple = false,
    required = false,
    withPortal,
    placeholder = t("Select a country"),
  } = props;

  const { data, loading } = useQuery<
    CountryPickerQuery,
    CountryPickerQueryVariables
  >(
    gql`
      query CountryPicker {
        countries {
          ...CountryPicker_country
        }
      }
      ${CountryPicker.fragments.country}
    `,
    { fetchPolicy: "cache-first" }
  );
  const [query, setQuery] = useState("");
  const debouncedQuery = useDebounce(query, 150);

  const options = useMemo(() => {
    const lowercaseQuery = debouncedQuery.toLowerCase();
    return (
      data?.countries?.filter((c) =>
        c.name.toLowerCase().includes(lowercaseQuery)
      ) ?? []
    );
  }, [data, debouncedQuery]);

  return (
    <Combobox<CountryPicker_CountryFragment>
      required={required}
      onChange={onChange}
      loading={loading}
      withPortal={withPortal}
      displayValue={(value) =>
        value
          ? ensureArray(value)
              .map((v) => v.name)
              .join(", ")
          : ""
      }
      by="code" /* FIXME: Upgrade @headlessui/react to > 1.6.6 to fix the comparison criteria */
      onInputChange={useCallback((event) => setQuery(event.target.value), [])}
      placeholder={placeholder}
      value={value}
      multiple={multiple}
      onClose={useCallback(() => setQuery(""), [])}
      disabled={disabled}
    >
      {options.map((option) => (
        <Combobox.CheckOption key={option.code} value={option}>
          <div className="flex items-center">
            <img
              loading="lazy"
              src={option.flag}
              className="sr-hidden mr-2"
              width={16}
              height={11}
              alt="Country Flag"
            />
            {option.name}
          </div>
        </Combobox.CheckOption>
      ))}
    </Combobox>
  );
};

CountryPicker.fragments = {
  country: gql`
    fragment CountryPicker_country on Country {
      code
      alpha3
      name
      flag
    }
  `,
};

export default CountryPicker;
