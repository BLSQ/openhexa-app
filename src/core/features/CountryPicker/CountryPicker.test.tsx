import CountryPicker from "./CountryPicker";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockedProvider } from "@apollo/client/testing";

describe("CountryPicker", () => {
  it("renders", async () => {
    const onChange = jest.fn();

    const { container } = render(
      <MockedProvider mocks={[]}>
        <CountryPicker onChange={onChange} value={null} />
      </MockedProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
