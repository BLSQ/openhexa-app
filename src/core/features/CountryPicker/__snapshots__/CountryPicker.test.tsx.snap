// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryPicker renders 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      class="form-input flex w-full items-center rounded-md border-gray-300 shadow-sm disabled:border-gray-300 focus-within:outline-none focus:ring-transparent focus-visible:border-blue-500 disabled:cursor-not-allowed  sm:text-sm hover:border-gray-400"
    >
      <div
        class="mr-1 flex flex-1 items-center truncate"
      >
        <input
          aria-autocomplete="list"
          aria-expanded="false"
          autocomplete="off"
          class="flex-1 placeholder-gray-600 placeholder-opacity-70 outline-none"
          data-headlessui-state=""
          data-testid="combobox-input"
          id="headlessui-combobox-input-6fbe024f-2316-4265-a6e8-d65a837e308a"
          placeholder="Select a country"
          role="combobox"
          type="text"
        />
      </div>
      <button
        aria-expanded="false"
        aria-haspopup="listbox"
        data-headlessui-state=""
        data-testid="combobox-button"
        id="headlessui-combobox-button-e67862f3-ccbf-4c51-b8ed-ed1d0420ea19"
        tabindex="-1"
        type="button"
      >
        <div
          class="ml-1 flex items-center gap-0.5 rounded-r-md text-gray-400 focus:outline-none"
        >
          <svg
            class="animate-spin h-5 w-5"
            data-testid="spinner"
            fill="none"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              class="opacity-25"
              cx="12"
              cy="12"
              r="10"
              stroke="currentColor"
              stroke-width="4"
            />
            <path
              class="opacity-75"
              d="M4 12a8 8 0 018-8V0C5.373 0 0 5.373 0 12h4zm2 5.291A7.962 7.962 0 014 12H0c0 3.042 1.135 5.824 3 7.938l3-2.647z"
              fill="currentColor"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</div>
`;
